node/cmd/spy/spy.go

```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
 	return nil
 }
 
+// TransactionIdMatches decodes both transactionIDs and checks if they are the same.
+func TransactionIdMatches(batch *vaa.BatchVAA, t *spyv1.BatchFilter) (matches bool, err error) {
+
+	// first check if the transaction IDs match
+	filterHash, err := vaa.StringToHash(t.TransactionId)
+	if err != nil {
+		return matches, status.Error(codes.InvalidArgument, fmt.Sprintf("failed to decode filter's txId: %v", err))
+	}
+
+	matches = filterHash == batch.TransactionID
+	return matches, nil
+}
+
+// BatchMatchFilter asserts that the obervation matches the values of the filter.
+func BatchMatchesFilter(batch *vaa.BatchVAA, f *spyv1.BatchFilter) (matches bool, err error) {
+
+	txMatch, err := TransactionIdMatches(batch, f)
+	if err != nil {
+		return matches, err
+	}
+
+	if txMatch {
+		// the BatchVAA's transaction ID matches the transaction ID of this filter.
+		// now check if the other properties of the filter match.
+
+		if obs := batch.Observations[0]; obs != nil {
+			obsVAA := obs.Observation
+
+			if obsVAA.EmitterChain == vaa.ChainID(f.ChainId) {
+				// the emitter chain of the observation matches the filter
+
+				if f.Nonce >= 1 {
```

Comment:
Thanks, I changed the signature of this function and simplified it a fair bit.

I left the `if f.Nonce >= 1`, context is, messages from the same transaction with the same nonce are batched together, with the exception of messages with nonce: `0`. It's a special case to allow developers to exclude messages from a batch. So because `0` has a special meaning, I opted to leave it `>= 1`